### DISCUSSION TITLE
Enter Pipe class between any of the existing ones

### DISCUSSION THREAD

**Discussion by KaloyanYosifov** [spatie/laravel-data#547] [Ideas]
Hey I have a feature request which I would like to discuss before I create a PR. 

--- 

So this idea comes from an issue I stumbled today while I was working on a project. We have a DTO which has a variable `type` that gets filled by an optional route parameter. 

```
class SomeDTO extends Data {
    public function __construct(
        #[FromRouteParameter('type')]
        public ?SomeEnum $type
    ) {
    
    }
}
```

This is working very well so far. But the issue  is that our team has agreed to a spec where we send `type` as `null` in the url (Don't ask me why :skull: ). Example: https://example.org/posts/category/null

Now in this case I would want the `type` to be null in the DTO, so I don't get any cast or validation errors. This is not the case so  I created myself a pipe which I pass to the pipeline that checks if `type` is a string `null` and if it is I just set it to `null` type.

Here is how this works:

```
class SomeDTO extends Data {
    public function __construct(
        #[FromRouteParameter('type')]
        public ?SomeEnum $type
    ) {
    
    }

    public static function pipeline(): DataPipeline
    {
        return DataPipeline::create()
            ->into(static::class)
            ->through(AuthorizedDataPipe::class)
            ->through(MapPropertiesDataPipe::class)
            ->through(FillRouteParameterPropertiesDataPipe::class)
            ->through(FixTypeRouteParameterPropertyDataPipe::class)
            ->through(ValidatePropertiesDataPipe::class)
            ->through(DefaultValuesDataPipe::class)
            ->through(CastPropertiesDataPipe::class);
    }
}

```

Notice how I add `FixTypeRouteParameterPropertyDataPipe::class` just after `FillRouteParameterPropertiesDataPipe::class`. 

---

Now with that story done :sweat_smile: . My whole problem with the above solution is that I have to copy the `pipeline` definition of the parent to add the pipe after `FillRoute`. I would like to propose a method on `DataPipeline` where we can choose where we can add a new Pipe  which allows to use the default pipes without copying. This also allows to use the new default pipes introduced into the library

As an example: 

```
    public static function pipeline(): DataPipeline
    {
        return parent::pipeline()
            ->throughAfter(FixTypeRouteParameterPropertyDataPipe::class, FillRouteParameterPropertiesDataPipe::class);
    }
```

or maybe have `DataPipeline` support setting pipes array with a setter and have a getter for the user to manipulate. 

Let me know if this proposal makes sense

**Comment by rubenvanassche** [comment]
In such a case I would create a new abstract data object which has the pipeline like you've described, then every data object could extend from that data object and would run the extra pipe.
